### PR TITLE
Issue 21: Batch processing to enable historical processing

### PR DESCRIPTION
**Pull request by philipjyoon**
Have not yet been tested as a lambda function - only been tested as a console binary running in a main loop.

### CODE REVIEW

File: lambdas/batch_process/batch_process_lambda.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+# Requires these 4 env variables
+_ENV_MOZART_IP = "MOZART_IP"
+_ENV_GRQ_IP = "GRQ_IP"
+_ENV_ENDPOINT = "ENDPOINT"
+_ENV_JOB_RELEASE = "JOB_RELEASE"
+
+for ev in [_ENV_MOZART_IP, _ENV_GRQ_IP, _ENV_ENDPOINT, _ENV_JOB_RELEASE]:
+    if ev not in os.environ:
+        raise RuntimeError("Need to specify %s in environment." % ev)
```
[the remark this answers is not in the document]
Comment (philipjyoon) [reply lambdas/batch_process/batch_process_lambda.py:29]: The current wording is how it is phrased in all other existing lambda functions such as the following. I suggest that we keep the phrasing consistent for now and then change to something better if/when we do a refactor of all lambda functions
https://github.com/nasa/opera-sds-lambdas/blob/develop/lambdas/data-subscriber-query/data_subscriber_query_lambda.py

File: lambdas/batch_process/batch_process_lambda.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+MOZART_IP = os.environ[_ENV_MOZART_IP]
+GRQ_IP = os.environ[_ENV_GRQ_IP]
+ENDPOINT = os.environ[_ENV_ENDPOINT]
+JOB_RELEASE = os.environ[_ENV_JOB_RELEASE]
+
+MOZART_URL = 'https://%s/mozart' % MOZART_IP
+JOB_SUBMIT_URL = "%s/api/v0.1/job/submit?enable_dedup=false" % MOZART_URL
```
[the remark this answers is not in the document]
Comment (philipjyoon) [reply lambdas/batch_process/batch_process_lambda.py:36]: Same comment as above. This is how it's coded up in other similar lambda functions. Suggest we keep consistency for now and refactor later as needed.

File: lambdas/batch_process/batch_process_lambda.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+ES_DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%S"
+ES_INDEX = 'batch_proc'
+LOGGER = logging.getLogger(ES_INDEX)
+eu = ElasticsearchUtility('http://%s:9200' % GRQ_IP, LOGGER)
```
[the remark this answers is not in the document]
Comment (philipjyoon) [reply lambdas/batch_process/batch_process_lambda.py:41]: Agreed. Will make it an ENV variable.

File: lambdas/batch_process/batch_process_lambda.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    if req.status_code != 200:
+        req.raise_for_status()
+    result = req.json()
+    print("Request Result: %s" % result)
+
+    if "result" in result.keys() and "success" in result.keys():
+        if result["success"] is True:
+            job_id = result["result"]
+            print("submitted job: %s job_id: %s" % (job_spec, job_id))
```
[the remark this answers is not in the document]
Comment (philipjyoon) [reply lambdas/batch_process/batch_process_lambda.py:84]: Same comment as the first two above. This is done the same way in all other existing lambda functions. Suggest we keep consistency for now and refactor later as needed.

To answer the direct question here, a successful run of the lambda will be evident when a job is submitted to the PCM and can be viewed in RabbitMQ and Figaro. This lambda runs independently of the PCM and does not / cannot return any data structure to any other component in the PCM system.

File: lambdas/batch_process/batch_process_lambda.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            continue
+
+        # update last_attempted_proc_data_date here
+        eu.update_document(id=doc_id,
+                           body={"doc_as_upsert": True,
+                                 "doc": {
+                                     "last_attempted_proc_data_date": e_date, }},
+                           index=ES_INDEX)
+
+        print("Submitting query job for", p.label, "with start date", s_date, "and end date", e_date)
```
[the remark this answers is not in the document]
Comment (philipjyoon) [reply lambdas/batch_process/batch_process_lambda.py:153]: Same as most comments above. This is how it's coded up in other similar lambda functions. Suggest we keep consistency for now and refactor later as needed.

I don't know if these end up in cloudwatch or not. If not and we want it to, we would want to do that for all lambda functions.

File: lambdas/batch_process/batch_process_lambda.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (philipjyoon) [reply lambdas/batch_process/batch_process_lambda.py]: This is done

**Comment by philipjyoon** [comment]
This branch must be used in conjunction with the issue_21 branch of the repo https://github.com/nasa/opera-sds-pcm

**Comment by philipjyoon** [comment]
The current version of hysds does not retain job tags beyond the very first one. So until this issue is resolved we are not able to assign a unique tag for historical/batch process jobs. One way to get around this may be to concatenate "historical/batch" to the existing job tags. This may confuse the operators - we'd have to discuss.

A Hysds core ticket has been created here:
https://hysds-core.atlassian.net/browse/HC-449

File: lambdas/batch_process/batch_process_lambda.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (riverma) [reply lambdas/batch_process/batch_process_lambda.py:30]: That’s fair. Sounds good @philipjyoon

File: lambdas/batch_process/batch_process_lambda.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+        # If it's not time to run yet, just continue
+        if new_last_run_date > now:
+            continue
+
+        # Update last_run_date here
+        eu.update_document(id=doc_id,
+                           body={"doc_as_upsert": True,
+                                 "doc": {
+                                     "last_run_date": now.strftime(ES_DATETIME_FORMAT), }},
```
[the remark this answers is not in the document]
Comment (riverma) [reply lambdas/batch_process/batch_process_lambda.py:118]: Interesting @philipjyoon - if the CLI does not allow for race conditions such as this, I suggest we mark this issue closed. I’m curious to hear how you’ve limited it through the CLI. Can you provide an explanation here? Don’t let this discussion stop you from merging though.

File: lambdas/batch_process/batch_process_lambda.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (riverma) [reply lambdas/batch_process/batch_process_lambda.py:155]: There’s probably enough minor usability improvements to warrant a ticket at this point. I agree let’s push this to later.

**Review by riverma — APPROVED** [review]
Great work @philipjyoon - this looks solid. I’m assuming you’ve tested this locally or on a test server? If so - go for merge.

**Comment by riverma** [comment]
> The current version of hysds does not retain job tags beyond the very first one. So until this issue is resolved we are not able to assign a unique tag for historical/batch process jobs. One way to get around this may be to concatenate "historical/batch" to the existing job tags. This may confuse the operators - we'd have to discuss.
> 
> A Hysds core ticket has been created here: https://hysds-core.atlassian.net/browse/HC-449

From the ticket you made – it seems like this is more of a UI issue on the HySDS side? If so, the patch to fix might be quite trivial. Would you mind taking a quick look at the HySDS core code base to see which block of code might enable this tag fix from a UI perspective? If we can reference that block of code or even submit a quick patch – we could get this taken care of pretty quickly with a point release of HySDS.

File: lambdas/batch_process/batch_process_lambda.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipjyoon) [reply lambdas/batch_process/batch_process_lambda.py:118]: @riverma The latest design of CLI does not allow for modification of a batch_proc at all. (This was desired behavior the last time the team had a discussion on this) In the latest design doc this command has been crossed out. So even if there were several operators working with batch_procs, they will not be able to edit any parameters of a batch_proc at all - they can only delete or create a new one.

**Comment by philipjyoon** [comment]
> From the ticket you made – it seems like this is more of a UI issue on the HySDS side? If so, the patch to fix might be quite trivial. Would you mind taking a quick look at the HySDS core code base to see which block of code might enable this tag fix from a UI perspective? If we can reference that block of code or even submit a quick patch – we could get this taken care of pretty quickly with a point release of HySDS.

I spoke with Gerald and it's most likely an API/orchestrator issue. The tags do not make it into the ES.

**Comment by riverma** [comment]
> > From the ticket you made – it seems like this is more of a UI issue on the HySDS side? If so, the patch to fix might be quite trivial. Would you mind taking a quick look at the HySDS core code base to see which block of code might enable this tag fix from a UI perspective? If we can reference that block of code or even submit a quick patch – we could get this taken care of pretty quickly with a point release of HySDS.
> 
> I spoke with Gerald and it's most likely an API/orchestrator issue. The tags do not make it into the ES.

Ah - okay I see. It's more complicated then. Thanks for filing the ticket. @hhlee445 - can [this ticket](https://hysds-core.atlassian.net/browse/HC-449) be discussed at your next tag-up with other PCM leads?

**Comment by philipjyoon** [comment]
Running the following command will result in one historical granule from Jan 1, 2015 to be queued for download because it intersects with North America:

python daac_data_subscriber.py query -p ASF --endpoint=OPS -c SENTINEL-1A_SLC -s 2015-01-01T00:00:00 -e 2015-01-01T01:00:00  --job-queue=opera-job_worker-slc_data_query --release-version=issue_21 --chunk-size=1 --north-america-only

**Pull request merged**